tests/test_hooks.py

```diff
@@ -0,0 +1,176 @@
+from datetime import datetime
+from typing import Dict, List
+
+import attr
+
+
+class TestTransformHook:
+    """Tests for `attrs(tranform_value_serializer=func)`"""
+
+    def test_hook_applied(self):
+        """
+        The transform hook must be applied to all attributes.  Types will not
+        be available here (they only work with type annotations).
+        """
+        results = []
+
+        def hook(cls, attribs):
+            results[:] = [a.name for a in attribs]
+            return attribs
+
+        @attr.s(field_transformer=hook)
+        class C:
+            x = attr.ib()
+            y = attr.ib(type=int)
+
+        assert results == ["x", "y"]
+
+    def test_hook_applied_auto_attrib(self):
+        """
+        The transform hook must be applied to all attributes and type
+        annotations must be detected.
+        """
+        results = []
+
+        def hook(cls, attribs):
+            results[:] = [(a.name, a.type) for a in attribs]
+            return attribs
+
+        @attr.s(auto_attribs=True, field_transformer=hook)
+        class C:
+            x: int
+            y: str = attr.ib()
+
+        assert results == [("x", int), ("y", str)]
+
+    def test_hook_applied_modify_attrib(self):
+        """The transform hook must be able to modify attributes."""
+
+        def hook(cls, attribs):
+            return [a._assoc(converter=a.type) for a in attribs]
```

Comment:
We should probably make `_assoc` public then? 🤔 Or at least document it?

Reply:
Why not both? 😼